### DISCUSSION TITLE
Offline audio

### DISCUSSION THREAD

**Discussion by gokadzev** [gokadzev/Musify#326] [Ideas]
Discussion about Offline audio feature

**This feature replaces 'download' feature, I have some thoughts about it so you may still have a possibility to copy your offline audios as files, will see...**

**Comment by gokadzev** [comment]
@foozaweb

There you go: https://github.com/gokadzev/Musify/tree/offline_audio

**Reply by gokadzev** [reply]
Issues (will add more):

- audio_service doesn't see or doesn't have access on it so it's not being show on the notificaton (nvm i think it works?)
- It also isn't shown in song bar or in player cubes (I think this one can be fixed)
- We need the same way for checking, removing and adding (in some places i give it whole song skeleton, in some places just song id)

**Reply by foozaweb** [reply]
checking now. Will revert shortly

**Reply by foozaweb** [reply]
Then for local songs

`Future<void> playLocalSong(int index) async {
  if (!currentModeIsLocal) {
    await _playlist.clear();
    await moveAudiosToQueue();
    await setNewPlaylist();
  }

  if (index < 0 || index >= _playlist.children.length) return;

 await audioHandler.skipToQueueItem(index); 
  await audioPlayer.play();
}`

should rather be this

`Future<void> playLocalSong(int index) async {
  if (!currentModeIsLocal) {
    await _playlist.clear();
    await moveAudiosToQueue();
    await setNewPlaylist();
  }

  if (index < 0 || index >= _playlist.children.length) return;

 await audioHandler.skipToQueueItem(index); //this is supposed to interface with Audio Handler, but it is returning initialization error
  audioPlayer.seek(Duration.zero, index: index); //The audioPlayer could directly skip to the song index
  await audioPlayer.play();
}`

**Reply by gokadzev** [reply]
What's that

**Reply by gokadzev** [reply]
It's from previous Musify version code which was created for supporting both: local and online playing. We don't need at all.

**Comment by foozaweb** [comment]
You are a genius bro, This is so awesome

**Reply by gokadzev** [reply]
Thank you 😄 but it needs more to be usable feature. I will continue working on it.

**Reply by foozaweb** [reply]
`Icon(
  userOfflineSongs.contains(song)
      ? FluentIcons.cellular_off_24_regular
      : FluentIcons.cellular_24_regular,
),
onPressed: () => userOfflineSongs.contains(song)
    ? removeSongFromOffline(song['ytid'])
    : makeSongOffline(song),`


I found the above useful for your icon in song_bar

**Reply by gokadzev** [reply]
The problem is that i modify it with adding additional properties, so it won't match.

**Reply by foozaweb** [reply]
If you mean the Icon? I tried the it though and seems to work fine

**Reply by gokadzev** [reply]
> If you mean the Icon? I tried the it though and seems to work fine

Did you pull the latest updates?

**Reply by foozaweb** [reply]
No. I just located the codes you edited

**Reply by gokadzev** [reply]
> `Icon( userOfflineSongs.contains(song) ? FluentIcons.cellular_off_24_regular : FluentIcons.cellular_24_regular, ), onPressed: () => userOfflineSongs.contains(song) ? removeSongFromOffline(song['ytid']) : makeSongOffline(song),`

Okay, I think it works.

**Reply by foozaweb** [reply]
You did a nice job here. Kudos
